python-api/tests/integrationtests/baseclass.py

```diff
@@ -0,0 +1,412 @@
+        """Increment Quit-Counter, if it reaches call_count,
+        Quit the MainLoop"""
+        self.quit_count += 1
+        if self.quit_count == call_count:
+            self.quit_mainloop()
+
+
+class IntegrationTestbaseCompare(IntegrationTestbase):
+    """Base class for integration tests that require image capture
+    and comparison. Used for Video- and Audio-Tests.
+    """
+
+    def expect_frame(self, filename, port, turns=5, timeout=0.5):
+        """Read frames from the server and compare them against filename.
+        Return when a match is found or timeout seconds have passed"""
+
+        filepath = os.path.join(
+            os.path.dirname(__file__),
+            'reference_frames',
+            filename)
+
+        self.log.debug("loading reference image %s", filepath)
+        expected = scipy.misc.imread(filepath).astype(float)
+
+        for turn in range(0, turns):
+            folder = tempfile.mkdtemp()
+
+            # build frame-fetching gstreamer pipeline
+            command = self.get_capture_command(
+                port, os.path.join(folder, "frame%d.png"))
+
+            self.log.info("running command for 0.5s (turn %u of %u)",
+                          turn, turns)
+            self.log.debug("%s", command)
+            proc = subprocess.Popen(shlex.split(command))
+            try:
+                proc.wait(timeout=timeout)
+                assert proc.poll() == 0
+            except subprocess.TimeoutExpired:
+                proc.kill()
+                proc.wait()
+
+            self.log.debug("command finished, comparing generated frames")
+            (success, img, diff) = self.compare_frames(expected, folder)
+
+            self.log.debug("removing temp-folder")
+            shutil.rmtree(folder)
+
+            if success:
+                break
+
+        if not success:
+            self.log.debug("last turn failed, "
+                           "saving images for human inspection")
+            self.save_images(filename, img, expected, diff)
+
+        assert success
+
+    def compare_frames(self, expected, folder):
+        """compare frames generated by the pipeline-command
+        against an expected image"""
+        for frame in sorted(glob(os.path.join(folder, "*.png"))):
+            # load generared image
+            img = scipy.misc.imread(frame).astype(float)
+
+            # Calulcate image-difference
+            diff = abs(img - expected)
+
+            # Calculate comparison-score
+            #   # Manhattan norm
+            #   m_norm = sum(abs(diff))
+            #
+            #   # Zero norm
+            #   z_norm = scipy.linalg.norm(diff.ravel(), 0) / img.size
+
+            # Maximal deviation between individual subpixels
+            max_dev = max(diff.ravel())
+
+            self.log.debug("comparison-results on frame %s: max_dev=%f",
+                           frame, max_dev)
+
+            # Compare calculated score
+            if max_dev <= 3:
```

Comment:
It was a guessed number. Changing it to 0 requires pixel-perfect matches which seem to work locally. We'll see what travis thinks about it.